File: python/src/lib/calculator/domain/experiment.py
# Copyright 2022 The Bucketeer Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from datetime import datetime, timedelta

from proto.experiment import experiment_pb2


class Experiment:
    def __init__(self, pb: experiment_pb2.Experiment):
        self.pb = pb

    def is_updated_to_running(self, now: datetime) -> bool:
        now_unix = int(now.timestamp())
        if self.pb.status == experiment_pb2.Experiment.WAITING:
            if self.pb.start_at <= now_unix:
                return True
        return False

    def is_updated_to_finish(self, now: datetime) -> bool:
        two_days_ago_unix = int((now - timedelta(days=2)).timestamp())
        if self.pb.status in [
            experiment_pb2.Experiment.WAITING,
            experiment_pb2.Experiment.RUNNING,
        ]:
            if self.pb.stop_at < two_days_ago_unix:
                return True
        return False

    def is_calculated(self, now: datetime) -> bool:
        now_unix = int(now.timestamp())
        if self.pb.start_at <= now_unix:
            return True
        return False

File: pkg/experiment/storage/v2/experiment.go
// Copyright 2026 The Bucketeer Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//go:generate mockgen -source=$GOFILE -package=mock -destination=./mock/$GOFILE
package v2

import (
	"context"
	"errors"

	pkgErr "github.com/bucketeer-io/bucketeer/v2/pkg/error"
	"github.com/bucketeer-io/bucketeer/v2/pkg/experiment/domain"
	proto "github.com/bucketeer-io/bucketeer/v2/proto/experiment"
)

var (
	ErrExperimentAlreadyExists = pkgErr.NewErrorAlreadyExists(
		pkgErr.ExperimentPackageName,
		"already exists",
	)
	ErrExperimentNotFound = pkgErr.NewErrorNotFound(
		pkgErr.ExperimentPackageName,
		"not found",
		"experiment",
	)
	ErrExperimentUnexpectedAffectedRows = pkgErr.NewErrorUnexpectedAffectedRows(
		pkgErr.ExperimentPackageName,
		"unexpected affected rows",
	)
	ErrExperimentCannotBeArchived = pkgErr.NewErrorInvalidArgNotMatchFormat(
		pkgErr.ExperimentPackageName,
		"cannot be archived",
		"experiment_status",
	)
)

// Shared list-query errors returned by ExperimentStorage / GoalStorage implementations.
var (
	ErrInvalidCursor  = errors.New("experiment/storage/v2: invalid cursor")
	ErrInvalidOrderBy = errors.New("experiment/storage/v2: invalid order by")
)

type ExperimentStorage interface {
	CreateExperiment(ctx context.Context, e *domain.Experiment, environmentId string) error
	UpdateExperiment(ctx context.Context, e *domain.Experiment, environmentId string) error
	GetExperiment(ctx context.Context, id, environmentId string) (*domain.Experiment, error)
	ListExperiments(
		ctx context.Context,
		params ListExperimentsParams,
	) ([]*proto.Experiment, int, int64, error)
	// GetExperimentSummary returns the total count of experiments by status.
	GetExperimentSummary(ctx context.Context, environmentID string) (*ExperimentSummary, error)
}

type ListExperimentsParams struct {
	EnvironmentID  string
	Archived       *bool
	FeatureID      string
	FeatureVersion *int32
	StartAt        int64
	StopAt         int64
	Maintainer     string
	Statuses       []proto.Experiment_Status
	SearchKeyword  string
	OrderBy        proto.ListExperimentsRequest_OrderBy
	OrderDirection proto.ListExperimentsRequest_OrderDirection
	PageSize       int
	Cursor         string
}

type ExperimentSummary struct {
	TotalWaitingCount int64
	TotalRunningCount int64
	TotalStoppedCount int64
}
